### PR TITLE
Rescale ODEs to match Euler solver

### LINKED ISSUE #15217: Euler ODE solver should recognize when there is a missing factor of x
```
>>> classify_ode(eqn.doit().expand())
()
>>> classify_ode((x**2*eqn.doit()).expand())
('nth_linear_euler_eq_nonhomogeneous_variation_of_parameters', 'nth_linear_euler_eq_nonhomogeneous_variation_of_parameters_Integral')
```

Probably the simplest way to fix this would be to make the ODE monic in the highest derivative and change the matching to check against that.

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### References to other Issues or PRs

Fixes #15217 

Needs a fix for #15433 for one of the tests

#### Brief description of what is fixed or changed

When matching an ODE against the Euler form this patch multiplies the coefficients by an appropriate power of `x`. This means that `dsolve` can recognise more ODEs as being Euler ODEs and can solve some ODEs that it couldn't before.

This example gives `NotImplementedError` on current master:
```julia
In [7]: eqn = -exp(x) + (x*Derivative(f(x), (x, 2)) + Derivative(f(x), x))/x

In [8]: eqn
Out[8]: 
           2                 
          d          d       
       x⋅───(f(x)) + ──(f(x))
           2         dx      
   x     dx                  
- ℯ  + ──────────────────────
                 x           

In [9]: dsolve(eqn, f(x))
Out[9]: 
                                            ⌠               
                                 x          ⎮    x          
f(x) = C₁ + C₂⋅log(x) + (x - 1)⋅ℯ ⋅log(x) - ⎮ x⋅ℯ ⋅log(x) dx
                                            ⌡    
```

#### Other comments

I've had to adjust the tests in `classify_ode` because the Euler solver now matches many trivial ODEs.

I found #15433 while working on this. One of the tests uses an awkward workaround for that. Probably #15433 should be fixed first.

#### Release Notes

* solvers
    * dsolve now rescales ODEs so that the Euler solver can be used in more cases

### CODE REVIEW

File: sympy/solvers/tests/test_ode.py [hunk 1]
```diff
@@ -2420,6 +2428,16 @@ def test_nth_order_linear_euler_eq_nonhomogeneous_variation_of_parameters():
     assert dsolve(eq, f(x), hint=our_hint).rhs in (sol, sols)
     assert checkodesol(eq, sol, order=2, solve_for_func=False)[0]
 
+    eq = -exp(x) + (x*Derivative(f(x), (x, 2)) + Derivative(f(x), x))/x
+    sol = Eq(f(x), C1 + C2*log(x) + (x - 1)*exp(x)*log(x) - Integral(x*exp(x)*log(x), x))
```
Comment (sidhantnagpal) [inline sympy/solvers/tests/test_ode.py]: The following will work without calling `.replace()`:
```
    eq = -exp(x) + (x*Derivative(f(x), (x, 2)) + Derivative(f(x), x))/x
    from sympy.integrals.risch import NonElementaryIntegral
    sol = Eq(f(x), C1 + C2*log(x) + (x - 1)*exp(x)*log(x) - NonElementaryIntegral(x*exp(x)*log(x), x))
    assert our_hint in classify_ode(eq)
    assert dsolve(eq, f(x), hint=our_hint) == sol
    assert checkodesol(eq, sol, order=2, solve_for_func=False)[0]
```

**Comment by oscarbenjamin** [comment]
I've just changed the test so that the expected solution is in terms of `NonElementaryIntegral`. I'm not totally satisfied with that test since it is actually possible to get that solution in a better form using `Ei(x)` but `integrate` can't do it yet: #15431 

Otherwise I'm happy with this PR now.

**Comment by asmeurer** [comment]
Why not make the ODE monic, as I suggested in the issue. That would also handle other things, like the whole ODE being multiplied by some function.

**Comment by oscarbenjamin** [comment]
I've changed it so that it makes the equation *almost* monic. It multiples through by `x**n/a` where a is the leading coefficient. This puts into something like the monic form of an Euler ODE.

I've also added a test for the case that the ODE is multiplied through by a function of `x`.

**Comment by asmeurer** [comment]
I see. I thought it was only detecting constant `a`.

**Comment by oscarbenjamin** [comment]
It was - I just pushed another commit before my previous comment.

**Pull request merged**